tikv/mvcc.go

```diff
@@ -144,6 +152,84 @@ func (store *MVCCStore) Prewrite(reqCtx *requestCtx, mutations []*kvrpcpb.Mutati
 	}
 
 	lockBatch := newWriteLockBatch(reqCtx)
+	// Check the DB.
+	txn := reqCtx.getDBReader().GetTxn()
+	keys := make([][]byte, len(mutations))
+	for i, m := range mutations {
+		keys[i] = m.Key
+	}
+	items, err := txn.MultiGet(keys)
+	if err != nil {
+		return []error{err}
+	}
+
+	for i, m := range mutations {
+		// Note the use of `forUpdateTS` here!
+		oldMeta, oldVal, err := store.checkConflictInDB(reqCtx, txn, items[i], forUpdateTS)
+		if err != nil {
+			anyError = true
+		}
+		errs[i] = err
+		if !anyError {
+			lock := mvcc.MvccLock{
+				MvccLockHdr: mvcc.MvccLockHdr{
+					StartTS:    startTS,
+					Op:         uint8(kvrpcpb.Op_PessimisticLock),
+					HasOldVer:  oldMeta != nil,
+					TTL:        uint32(ttl),
+					PrimaryLen: uint16(len(primary)),
+				},
+				Primary: primary,
+				Value:   m.Value,
+				OldMeta: oldMeta,
+				OldVal:  oldVal,
+			}
+			lockBatch.set(m.Key, lock.MarshalBinary())
+		}
+	}
+	if anyError {
+		return errs
+	}
+	err = store.writeLocks(lockBatch)
+	if err != nil {
+		return []error{err}
+	}
+	return nil
+}
+
+func (store *MVCCStore) Prewrite(reqCtx *requestCtx, mutations []*kvrpcpb.Mutation, primary []byte, startTS uint64, ttl uint64) []error {
+	regCtx := reqCtx.regCtx
+	hashVals := mutationsToHashVals(mutations)
+	errs := make([]error, 0, len(mutations))
+	anyError := false
+
+	regCtx.acquireLatches(hashVals)
+	defer regCtx.releaseLatches(hashVals)
+	lockBatch := newWriteLockBatch(reqCtx)
+	var hasPessimistic bool
+
+	// Must check the LockStore first.
+	for _, m := range mutations {
+		lock, err := store.checkConflictInLockStore(reqCtx, m, startTS)
+		if err != nil {
+			anyError = true
+		}
+		if lock != nil {
+			if lock.Op == uint8(kvrpcpb.Op_PessimisticLock) {
+				// lockstore doesn't support update for now, delete the lock first.
+				lockBatch.delete(m.Key)
+				hasPessimistic = true
```

Comment:
maybe we should only set `hasPessimistic` flag just for in current `m`, other mutators also need check conflict?

Reply:
pessimistic lock never conflicts in prewrite.